### DISCUSSION TITLE
Match exporters interface with the opentelemetry-go SDK?

### DISCUSSION THREAD

**Discussion by dmathieu** [open-telemetry/opentelemetry-collector#2416] [Ideas]
Hey,

I'm not opening this as an issue, as it may have been covered elsewhere, and I feel I'm missing something regarding the reason this isn't the case.
It would be very convenient to be able to use the same exporter in both a non-collector project, and one transmitting its data to a collector.

So, have you considered matching the exporters' interfaces to match exporters from opentelemetry-go SDK?

**Comment by bogdandrutu** [comment]
Hello @dmathieu, I thought about that, but unfortunately there are limitations about the SDK go (cannot depend on proto) which makes things a bit harder. I think a small library "to/from otel-go representation" is a good starting point. Also the collector wants to develop the internal representation in a way that will be as performant as possible for cases where otlp is received and exported, which means we will at one point add capabilities like lazy unmarshaling, etc.

**Comment by kinthaiofficial** [comment]
The interface alignment question between collector exporters and the Go SDK is something we've thought about a lot when instrumenting agent-to-agent communication.

The mismatch becomes painful when you want to trace delegation chains — e.g., Agent A delegates to Agent B which calls Agent C — and you need the spans to propagate correctly across process boundaries:

```go
// Exporter interface in collector vs SDK differ in push/pull model
// Collector: receiver pulls from pipeline
// SDK: SDK pushes to exporter

// Bridge pattern to unify them:
type CollectorBridge struct {
    sdkExporter sdktrace.SpanExporter
    mu          sync.Mutex
    pending     []sdktrace.ReadOnlySpan
}

func (b *CollectorBridge) Export(ctx context.Context, req pdata.Traces) error {
    spans := convertToSDKSpans(req)
    return b.sdkExporter.ExportSpans(ctx, spans)
}
```

For multi-agent tracing we ended up keeping them separate — SDK for in-process, collector pipeline for cross-agent — because trying to unify created more friction than it solved. The collector's batching and retry logic is hard to replicate in the SDK interface.

We wrote about tracing delegation chains in distributed agent systems here: https://blog.kinthai.ai/221-agents-multi-agent-coordination-lessons
